lib/std/fs/Dir.zig

```diff
@@ -1740,16 +1740,23 @@ pub fn readLinkW(self: Dir, sub_path_w: []const u16, buffer: []u8) ![]u8 {
     return std.os.windows.ReadLink(self.fd, sub_path_w, buffer);
 }
 
-/// Read all of file contents using a preallocated buffer.
-/// The returned slice has the same pointer as `buffer`. If the length matches `buffer.len`
-/// the situation is ambiguous. It could either mean that the entire file was read, and
-/// it exactly fits the buffer, or it could mean the buffer was not big enough for the
-/// entire file.
+/// Read the entirety of the file at `file_path` into `buffer`.
+/// The returned slice has the same pointer as `buffer`. If the entirety of the file does not fit
+/// into `buffer`, `error.Truncated` is returned, and `buffer` contains the first `buffer.len` bytes
+/// of the file.
 pub fn readFile(self: Dir, file_path: []const u8, buffer: []u8) ![]u8 {
     var file = try self.openFile(file_path, .{});
     defer file.close();
 
     const end_index = try file.readAll(buffer);
+
+    if (end_index == buffer.len) {
+        var unused: [1]u8 = undefined;
+        if (try file.read(&unused) > 0) {
```

Comment:
This isn't the right thing to do for frame delimited formats (such as a file descriptor hooked up to an SCTP socket)

Reply:
You're correct, but I'm not aware of any way a `SOCK_SEQPACKET` or `SOCK_DGRAM` socket could possibly be opened with `openFile`. Feel free to correct me with a real-world proof-of-concept though.

Reply:
> Feel free to correct me with a real-world proof-of-concept though.

As a trivial but not improbable example: `openFile` on `/proc/self/fd/N`

> using `readAll` would be invalid for non-stream sockets anyway

Howso? I'd expect `readAll` on an SOCK_SEQPACKET to return the entire current packet (but no more).

Reply:
> As a trivial but not improbable example: `openFile` on `/proc/self/fd/N`

It is not possible to open such a fd with `open`/`openat`, here is a demonstration of this fact:
```ts
const std = @import("std");

pub fn main() !void {
    var child = std.ChildProcess.init(&.{"ncat", "-vvvv", "--sctp", "-l", "-p", "6969"}, std.heap.c_allocator);
    defer _ = child.kill() catch {};

    try child.spawn();

    std.time.sleep(200 * std.time.ns_per_ms);

    const sock = try std.os.socket(std.os.AF.INET, std.os.SOCK.SEQPACKET, std.os.IPPROTO.SCTP);

    const sa = try std.net.Address.parseIp4("127.0.0.1", 6969);

    // set default address
    try std.os.connect(sock, @ptrCast(&sa.in), sa.getOsSockLen());


    var d = try std.fs.openDirAbsolute("/proc/self/fd/", .{});
    defer d.close();

    var b: [16]u8 = undefined;
    const sock_str = try std.fmt.bufPrint(&b, "{d}", .{sock});

    // must fail
    _ = d.openFile(sock_str, .{}) catch |err| std.debug.print("{}\n", .{err});
}
```

```
$ zig build-exe readfile_procself.zig -lc
$ ./readfile_procself 
Ncat: Version 7.94 ( https://nmap.org/ncat )
NCAT DEBUG: Initialized fdlist with 103 maxfds
Ncat: Listening on [::]:6969
NCAT DEBUG: Added fd 3 to list, nfds 1, maxfd 3
Ncat: Listening on 0.0.0.0:6969
NCAT DEBUG: Added fd 4 to list, nfds 2, maxfd 4
NCAT DEBUG: Added fd 0 to list, nfds 3, maxfd 4
NCAT DEBUG: Initialized fdlist with 100 maxfds
NCAT DEBUG: selecting, fdmax 4
NCAT DEBUG: unexpected errno: 6
select returned 1 fds ready
NCAT DEBUG: fd 4 is ready
NCAT DEBUG: Swapping fd[0] (3) with fd[2] (0)
NCAT DEBUG: Removed fd 3 from list, nfds 2, maxfd 4
NCAT DEBUG: Swapping fd[1] (4) with fd[1] (4)
NCAT DEBUG: Removed fd 4 from list, nfds 1, maxfd 0
Ncat: Connection from 127.0.0.1:60238.
NCAT DEBUG: Added fd 5 to list, nfds 2, maxfd 5
NCAT DEBUG: Added fd 5 to list, nfds 1, maxfd 5
NCAT DEBUG: selecting, fdmax 5
/home/user/Downloads/zig-linux-x86_64-0.12.0-dev.2668+ddcea2cad/lib/std/debug.zig:188:31: 0x106549c in dumpCurrentStackTrace (readfile_procself)
        writeCurrentStackTrace(stderr, debug_info, io.tty.detectConfig(io.getStdErr()), start_addr) catch |err| {
                              ^
/home/user/Downloads/zig-linux-x86_64-0.12.0-dev.2668+ddcea2cad/lib/std/os.zig:5831:40: 0x102b578 in unexpectedErrno (readfile_procself)
        std.debug.dumpCurrentStackTrace(null);
                                       ^
/home/user/Downloads/zig-linux-x86_64-0.12.0-dev.2668+ddcea2cad/lib/std/os.zig:1900:49: 0x106a4ec in openatZ (readfile_procself)
            else => |err| return unexpectedErrno(err),
                                                ^
/home/user/Downloads/zig-linux-x86_64-0.12.0-dev.2668+ddcea2cad/lib/std/fs/Dir.zig:843:26: 0x102d406 in openFileZ (readfile_procself)
        try posix.openatZ(self.fd, sub_path, os_flags, 0);
                         ^
/home/user/Downloads/zig-linux-x86_64-0.12.0-dev.2668+ddcea2cad/lib/std/fs/Dir.zig:775:26: 0x1028100 in openFile (readfile_procself)
    return self.openFileZ(&path_c, flags);
                         ^
/home/user/src/random/readfile_procself.zig:26:19: 0x10270a1 in main (readfile_procself)
    _ = d.openFile(sock_str, .{}) catch |err| std.debug.print("{}\n", .{err});
                  ^
/home/user/Downloads/zig-linux-x86_64-0.12.0-dev.2668+ddcea2cad/lib/std/start.zig:585:37: 0x1028706 in main (readfile_procself)
            const result = root.main() catch |err| {
                                    ^
../sysdeps/nptl/libc_start_call_main.h:58:16: 0x7fa6c3c9c6c9 in __libc_start_call_main (../sysdeps/x86/libc-start.c)
../csu/libc-start.c:360:3: 0x7fa6c3c9c784 in __libc_start_main_impl (../sysdeps/x86/libc-start.c)
???:?:?: 0x1025e90 in ??? (???)
???:?:?: 0x0 in ??? (???)
error.Unexpected
```

> Howso? I'd expect `readAll` on an SOCK_SEQPACKET to return the entire current packet (but no more).

`readAll` relies on the file being a stream, and reads until either the buffer is full or `read` returns 0:

https://github.com/ziglang/zig/blob/0c725a354a801c84100011db5eb53ae6c58086c9/lib/std/fs/File.zig#L1038-L1048